### PR TITLE
Rjwebb/more cw fixes march 2024

### PR DESCRIPTION
**Pull request by rjwebb**
This PR has some more fixes that came out of working on Commonwealth:

- the `AbstractSigner` class in `SIWESigner` (this class represents the extension/wallet object being passed into `SIWESigner`, e.g. an injected MetaMask wallet or an ephemeral ethers wallet) should be more permissive so that we don't get a type error with Keplr
- the `getAddress` on `SolanaSigner` should return the address with the `solana:` prefix - this was a typo
- the default chain id value for `SolanaSigner` should be the genesis hash, not prefixed with `solana:`
- the keplr ethereum login method in the chat app should create a session when the connect button is clicked
- expect cosmos ethereum signers to return 0x signatures and addresses

Add more unit tests for the above:
- [x] signers test for cosmos ethereum signer
- [x] assert that `getAddress` returns something of the form `"<chain base>:<chain id>:<wallet address>"`

## How has this been tested?

- [x] CI tests pass
- [x] Tested with example-chat (including login, all signers, and exchanging messages)
- [ ] Tested with `@canvas-js/test-network`: (optional)

## Does this contain any breaking changes to external interfaces?

- [ ] Contract interfaces
- [ ] Core interface
- [ ] CLI
- [ ] Data storage formats, including IndexedDB, SQLite, or filesystem storage (will this break existing apps?)

### CODE REVIEW

File: packages/signatures/src/AbstractSessionSigner.ts [hunk 1]
```diff
@@ -103,9 +100,7 @@ export abstract class AbstractSessionSigner<AuthorizationData> implements Sessio
 			const { signer, session } = this.getCachedSession(message.topic, message.payload.address) ?? {}
 			assert(signer !== undefined && session !== undefined)
 
-			// only sign our own current sessions
-			// use a deep comparison
-			assert(deepEquals(message.payload, session))
+			assert(equals(json.encode(message.payload), json.encode(session)))
```
Comment (raykyri) [inline packages/signatures/src/AbstractSessionSigner.ts:103]: There was a reason I switched from comparing `json.encode` to our own `deepEquals` implementation - iirc there was a test that was failing with json.encode comparison; what's the reason for switching back?

File: packages/chain-cosmos/src/external_signers/ethereum.ts [hunk 2]
```diff
@@ -30,11 +30,15 @@ export const createEthereumSigner = (signer: EthereumSigner, bech32Prefix: strin
...
 		const encodedMessage = encodeReadableEthereumMessage(cosmosMessage)
-
 		const ethAddress = `0x${bytesToHex(fromBech32(signerAddress).data)}`
-		const hexSignature = await signer.signEthereum(chainId, ethAddress, encodedMessage)
+		let signature = await signer.signEthereum(chainId, ethAddress, encodedMessage)
+		// some signers prepend the `0x` in the signature result (metamask) and some don't (keplr)
+		// we want to return the signature without this 0x prefix
+		if (signature.slice(0, 2) === "0x") {
+			signature = signature.slice(2)
+		}
```
Comment (raykyri) [inline packages/chain-cosmos/src/external_signers/ethereum.ts:39]: wow

Comment (rjwebb) [reply packages/chain-cosmos/src/external_signers/ethereum.ts:39]: ah wait i meant to get rid of this - i think we should just force the `signEthereum` function to return a signature with `0x...`

**Comment by raykyri** [comment]
basically looks good to me so far, seems like you've already flagged the issue with the near signer

there might be a case for sticking with deepEquals and unit testing it. i can write unit tests for that if you're busy with commonwealth stuff

File: packages/signatures/src/AbstractSessionSigner.ts (hunk 1 above)
Comment (rjwebb) [reply packages/signatures/src/AbstractSessionSigner.ts:103]: oh sorry, i misread the git commit log and thought i had added it

the `deepEquals` function fails on Uint8Arrays:

```
> let a = new Uint8Array([0,1,2])
undefined
> let b = new Uint8Array([0,1,2])
undefined
> deepEquals(a,b)
false
```

the reason why the interface tests pass is that we don't have the test case where we have two sessions that have the same data but are different objects in memory

Comment (rjwebb) [reply packages/signatures/src/AbstractSessionSigner.ts:103]: I'm happy to just fix the deepEquals function to make this work and add another test case, arguably it should just have its own unit test

Comment (raykyri) [reply packages/signatures/src/AbstractSessionSigner.ts:103]: oh gotcha. let me fix that

**Pull request merged**